Hazelcast Azure Integration Improvements(see description)

alparslanavci opened the pull request:
With this update:
- The Azure Java SDK usage is replaced by the [Azure REST API](https://docs.microsoft.com/en-us/rest/api/azure/).
- Added support for [virtual machine scale sets ](https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/overview).
- Made tag setting optional, the plugin can work without setting any tags now.
- Upgraded Hazelcast version from 3.x to 4.0.

Many thanks to @gokhanoner for his contributions in this update.

Fixes #10, fixes #11, fixes #12, fixes #17.

mesutcelik commented:
I see this warning when I do `mvn clean install`

```
WARNING: Name of the hazelcast schema location[hazelcast-config-3.6.xsd] is incorrect, using default
```
3.6?

I also see that exception but no test failure

```
Running com.hazelcast.azure.AzureDiscoveryStrategyTest
Feb 07, 2020 11:04:53 AM com.hazelcast.azure.AzureDiscoveryStrategy
WARNING: Cannot discover nodes, returning empty list
java.lang.RuntimeException: Error while checking Azure instances
	at com.hazelcast.azure.AzureDiscoveryStrategy.discoverNodes(AzureDiscoveryStrategy.java:137)
	at com.hazelcast.azure.AzureDiscoveryStrategyTest.discoverNodesException(AzureDiscoveryStrategyTest.java:153)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.mockito.internal.runners.JUnit45AndHigherRunnerImpl.run(JUnit45AndHigherRunnerImpl.java:37)
	at org.mockito.runners.MockitoJUnitRunner.run(MockitoJUnitRunner.java:62)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

mesutcelik commented on checkstyle/ClassHeader.txt:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
```
Licensing changed in 4.0-compatiblity so you need to update your branch too.

alparslanavci replied:
done 👍

mesutcelik commented on pom.xml:
```diff
@@ -1,35 +1,39 @@
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+
+        <!-- we want to remain java 1.6 compatible -->
+        <jdk.version>1.6</jdk.version>
```
1.8?

alparslanavci replied:
done 👍

mesutcelik commented on pom.xml:
```diff
@@ -62,301 +66,147 @@
+            <id>gokhanoner</id>
+            <name>Gokhan Oner</name>
+            <email>gokhan@hazelcast.com</email>
+        </developer>
```
you need to add your name too.

alparslanavci replied:
I think it is better to keep only community contributors here.

mesutcelik commented on pom.xml:
```diff
@@ -1,35 +1,39 @@
     <artifactId>hazelcast-azure</artifactId>
-    <version>1.2.3-SNAPSHOT</version>
-    <url>http://www.hazelcast.com/</url>
+    <version>1.3.0-SNAPSHOT</version>
```
I assume we are aiming 2.0 so 2.0-SNAPSHOT needs to be put for now.

alparslanavci replied:
done 👍

mesutcelik commented on README.md:
```diff
@@ -69,8 +69,9 @@ In your Hazelcast configuration, use the `AzureDiscoveryStrategy` as shown below
+        <resource-group>RESOURCE-GROUP-NAME</resource-group>
+        <scale-set>SCALE-SET-NAME</scale-set>
+        <tag>TAG-NAME=HZLCAST001</tag>
       </azure>
```
Can we also add yaml example?

alparslanavci replied:
done 👍

mesutcelik commented on README.md:
```diff
@@ -81,16 +82,21 @@ You will need to setup [Azure Active Directory Service Principal credentials](ht
-- `group-name` - The Azure [resource group](https://azure.microsoft.com/en-us/documentation/articles/resource-group-portal/) name of the cluster. You can find this in the Azure [portal](https://portal.azure.com) or [CLI](https://npmjs.org/azure-cli).
+- `client-id` - The Azure Active Directory Service Principal client ID.
+- `client-secret` - The Azure Active Directory Service Principal client secret.
+- `tenant-id` - The Azure Active Directory tenant ID.
```
Aren't those parameters optional too?

alparslanavci replied:
done with a note 👍

leszko commented on pom.xml:
```diff
@@ -505,21 +356,14 @@
             <id>master</id>
             <properties>
-                <hazelcast.version>4.0-SNAPSHOT</hazelcast.version>
+                <hazelcast.version>3.12-SNAPSHOT</hazelcast.version>
```
Shouldn't the master be `4.1-SNAPSHOT`?

alparslanavci replied:
updated 👍

alparslanavci commented:
@mesutcelik, that exception is thrown on purpose to test how plugin behaves when it couldn't get any addresses from REST API.

alparslanavci commented on README.md:
```diff
@@ -11,7 +11,7 @@
 # Hazelcast Discovery Plugin for Microsoft Azure
 
-This project provides a discovery strategy for Hazelcast 3.6-RC+1 enabled applications running on Azure. It will provide all Hazelcast instances by returning VMs within your Azure resource group that are tagged with a specified value.
+This project provides a discovery strategy for Hazelcast 3.6-RC+1 enabled applications running on Azure. It will provide all Hazelcast instances by returning VMs within your Azure resource group. It supports virtual machine scale sets and tagging.
```
updated 👍 I had kept it to mention that the plugin is available from Hazelcast 3.6+, but it is better to change it to 4.0+.

alparslanavci commented on README.md:
```diff
@@ -11,7 +11,7 @@
-This project provides a discovery strategy for Hazelcast 3.6-RC+1 enabled applications running on Azure. It will provide all Hazelcast instances by returning VMs within your Azure resource group that are tagged with a specified value.
+This project provides a discovery strategy for Hazelcast 3.6-RC+1 enabled applications running on Azure. It will provide all Hazelcast instances by returning VMs within your Azure resource group. It supports virtual machine scale sets and tagging.
 
 ![Architectual diagram](img/azurespi.png)
```
I think the image still displays the generic structure. Which part of the image make you to think not convenient anymore?

alparslanavci commented on README.md:
```diff
@@ -69,8 +69,9 @@ In your Hazelcast configuration, use the `AzureDiscoveryStrategy` as shown below
         <subscription-id>SUB_ID</subscription-id>
-        <cluster-id>HZLCAST001</cluster-id>
-        <group-name>GROUP-NAME</group-name>
+        <resource-group>RESOURCE-GROUP-NAME</resource-group>
```
added 👍

alparslanavci commented on pom.xml:
```diff
@@ -62,301 +66,147 @@
                 <picUrl>https://avatars.githubusercontent.com/u/3053263?v=3</picUrl>
             </properties>
         </developer>
+        <developer>
```
I will keep this section since the plugin initially developed by the community.

alparslanavci commented:
@leszko please find the answers below:

1. Updated. Please check if we can make it more simpler.
2. Yes, we also have this feature in this plugin. I tried to mention it in the readme, but it looks it is not clear. I added a few more lines as you suggested. 
3. `4.0-compatibilty` branch is irrelevant as now, since all the codebase is updated. This PR is already compatible with Hazelcast 4.0.

leszko commented on src/main/java/com/hazelcast/azure/Tag.java:
```diff
@@ -0,0 +1,81 @@
+    Tag(String spec) {
+        String[] labelParts = spec.trim().split("\\s*=\\s*");
+        if (labelParts.length != 2) {
+            throw new IllegalArgumentException(String.format("Invalid tag specification: '%s'", spec));
```
Maybe `InvalidConfigurationException`? I think that is more clear to the end user.

alparslanavci replied:
Changed 👍

leszko commented on src/test/java/com/hazelcast/azure/AzureComputeApiTest.java:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    private String instancesResponseForPublicIPAddresses() {
+        return String.format("{\n" + "  \"value\": [\n" + "    {\n" + "      \"name\": \"ip02\",\n"
```
Could you format these string, so that they would be more readable? Like [here](https://github.com/hazelcast/hazelcast-kubernetes/blob/master/src/test/java/com/hazelcast/kubernetes/KubernetesClientTest.java#L67).

IntelliJ has a very good support for this, so it will do it automatically for you.

alparslanavci replied:
Done 👍

leszko commented on src/test/java/com/hazelcast/azure/AzureComputeApiTest.java:
```diff
@@ -0,0 +1,207 @@
+        assertEquals(expected, result);
+    }
+    /**
+     * Reponse recorded from the real Cloud Compute API call.
```
```suggestion
     * Response recorded from the real Cloud Compute API call.
```

leszko commented on src/test/java/com/hazelcast/azure/AzureMetadataApiTest.java:
```diff
@@ -0,0 +1,168 @@
+                + "  \"resource\": \"https://management.azure.com/\",\n" + "  \"token_type\": \"Bearer\"\n" + "}", ACCESS_TOKEN);
+    }
+
+    private String metadataResponse() {
```
Could you format the String to be more readable, like [here](https://github.com/hazelcast/hazelcast-kubernetes/blob/master/src/test/java/com/hazelcast/kubernetes/KubernetesClientTest.java#L67)?

alparslanavci replied:
Done 👍

mesutcelik commented on README.md:
```diff
@@ -69,28 +70,78 @@ In your Hazelcast configuration, use the `AzureDiscoveryStrategy` as shown below
+      multicast:
+        enabled: false
+      azure:
+        enabled: true
```
I think we should somehow stop explicitly disabling multicast. @mmedenjak Do we have a related github issue for that?

mmedenjak replied:
About having multicast disabled by default? No, we don't.

alparslanavci replied:
@mmedenjak not disabling multicast by default, but maybe automatically disabling multicast if any other join method is enabled?

mmedenjak replied:
No. There are several other places where one config excludes the other (think setting `className` vs `implementation` on various user-provided implementations, e.g. of map loader) and as a general rule we don't disable/set to `null` whenever the other is configured.
Generally speaking, we fail-fast instead of trying to be smart as we are never sure what the user really wanted to achieve with that configuration, maybe they actually wanted to enable both multicast and tcp-ip.

mesutcelik replied:
The problem is that the following config does not work without explicitly disabling multicast.
```
hazelcast:
  network:
    join:
      azure:
        enabled: true
```

mmedenjak replied:
I know. But overall, we prefer explicitness over magic.

alparslanavci commented on README.md:
```diff
@@ -69,28 +70,78 @@ In your Hazelcast configuration, use the `AzureDiscoveryStrategy` as shown below
+## Minimal Configuration Example
 
-Read more about how you can [tag your virtual machines](https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-tagging-arm/).
+If you setup your Azure resource identities with the correct access roles, you will need just the following minimal configuration to setup Hazelcast Azure Discovery:
```
Added 👍

mesutcelik replied:
I suggest to put `az` specific commands into a getting started guide instead of readme. The guide can be referenced from README. 

I would also put this Exception into a troubleshooting section or you can mention it while you describe Managed Identities.

alparslanavci replied:
Yes, I will remove `az` commands for now, since I found some more issues regarding them. I will work on it later and create a guide. We can give it a reference from README.

I will add info about the exception 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+        this.clientId = azureConfig.getClientId();
+        this.clientSecret = azureConfig.getClientSecret();
+        this.hasSMIRight = hasSMIRight(tenantId, clientId, clientSecret);
+        this.subscriptionId = subscriptionIdFromConfigOrMetadataApi(azureConfig);
```
I think the scenario is what we expect from a Hazelcast instance startup, which is fail-fast. If the user configures the instance with Azure Discovery without setting up the optional parameters and Azure API fails, then it is best to fail and not start the Hazelcast instance. 

Also, `AzureDiscoveryStrategy.start()` is called in the `Node` start, which is called in `HazelcastInstanceImpl` constructor. If an exception is thrown in `AzureDiscoveryStrategy.start()`, then the node fails to start.

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureProperties.java:
```diff
@@ -1,143 +1,97 @@
-    public static final PropertyDefinition CLIENT_ID = property("client-id", STRING);
+enum AzureProperties {
 
+    CLIENT("client", BOOLEAN, true, false),
```
Removed 👍 It was only used when retrieving resource group and scale-set, but client might also be run in the same Azure resource group or scale-set.

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureProperties.java:
```diff
@@ -1,143 +1,97 @@
-     * @param key the key to use for the property
-     * @param typeConverter the PropertyTypeConverter to convert the property
-     * @return PropertyDefition the PropertyDefition for they key
+     * The constant CLIENT_SECRET.
```
Updated 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java:
```diff
@@ -49,156 +52,99 @@
-                }
-                int port = Integer.parseInt(tags.get(clusterId));
-                DiscoveryNode node = buildDiscoveredNode(faultDomainId, vm, port);
+    private static DiscoveryNode createDiscoveryNode(AzureAddress azureAddress, int port)
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+ * Responsible for fetching the discovery information from Azure APIs.
+ */
+class AzureClient {
+    private static final Logger LOGGER = Logger.getLogger(AzureDiscoveryStrategy.class.getSimpleName());
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+        this.tag = azureConfig.getTag();
+    }
+
+    private boolean hasSMIRight(String tenantId, String clientId, String clientSecret) {
```
Fixed 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+    private final AzureComputeApi azureComputeApi;
+    private final AzureAuthenticator azureAuthenticator;
+
+    private final String tenantId;
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+            return azureConfig.getSubscriptionId();
+        }
+        LOGGER.finest("Property 'subscriptionId' not configured, fetching from the VM metadata service");
+        return RetryUtils.retry(new Callable<String>() {
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    String getAvailabilityZone() {
```
Added 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureMetadataApi.java:
```diff
@@ -0,0 +1,110 @@
+        this.metadata = metadata;
+    }
+
+    private void fillMetadata() {
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureAuthenticator.java:
```diff
@@ -0,0 +1,70 @@
+    String refreshAccessToken(String tenantId, String clientId, String clientSecret) {
+        try {
+            String extractAccessToken = callService(urlFor(tenantId), body(clientId, clientSecret));
+            return extractAccessToken(extractAccessToken);
```
Changed 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureComputeApi.java:
```diff
@@ -0,0 +1,159 @@
+        this.endpoint = endpoint;
+    }
+
+    private static JsonObject toJsonObject(JsonValue jsonValue) {
```
Done 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureComputeApi.java:
```diff
@@ -0,0 +1,159 @@
+        Map<String, AzureNetworkInterface> interfaces = new HashMap<String, AzureNetworkInterface>();
+
+        for (JsonValue item : toJsonArray(Json.parse(response).asObject().get("value"))) {
+            String id = item.asObject().getString("id", null);
```
Removed 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/AzureComputeApi.java:
```diff
@@ -0,0 +1,159 @@
+
+        for (JsonValue item : toJsonArray(Json.parse(response).asObject().get("value"))) {
+            String id = item.asObject().getString("id", null);
+            Set<Tag> tagList = new HashSet<Tag>(1);
```
Removed 👍

alparslanavci commented on src/main/java/com/hazelcast/azure/Utils.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    static boolean isBlank(final CharSequence cs) {
+        int strLen = cs == null ? 0 : cs.length();
```
Changed 👍

mesutcelik commented on src/main/java/com/hazelcast/azure/AzureAddress.java:
```diff
@@ -0,0 +1,69 @@
+package com.hazelcast.azure;
+
+/**
+ * Structure that represents the discovery output information.
```
We can be more specific here. It just represents public/private IP Addresses.

alparslanavci replied:
Updated 👍

mesutcelik commented on src/main/java/com/hazelcast/azure/AzureClient.java:
```diff
@@ -0,0 +1,121 @@
+        this.subscriptionId = subscriptionIdFromConfigOrMetadataApi(azureConfig);
+        this.resourceGroup = resourceGroupFromConfigOrMetadataApi(azureConfig);
+        this.scaleSet = scaleSetFromConfigOrMetadataApi(azureConfig);
+        this.tag = azureConfig.getTag();
```
I think we need isBlank check for tag?

alparslanavci replied:
The tag is optional, and can be blank. We check if it is set in here: https://github.com/alparslanavci/hazelcast-azure/blob/convert-rest-api/src/main/java/com/hazelcast/azure/AzureComputeApi.java#L74

mesutcelik commented on src/main/java/com/hazelcast/azure/AzureComputeApi.java:
```diff
@@ -0,0 +1,158 @@
+
+    Collection<AzureAddress> instances(String subscriptionId, String resourceGroup, String scaleSet,
+                                       Tag tag, String accessToken) {
+        String privateIpResponse = RestClient
```
Can you check if you can pass tag to be filtered on MetadataService instead of looping through networkInterfaces?

alparslanavci replied:
I already checked, unfortunately not much options on Azure API.

mesutcelik commented on README.md:
```diff
@@ -69,28 +68,131 @@ In your Hazelcast configuration, use the `AzureDiscoveryStrategy` as shown below
+- `subscription-id` - *(Optional)* The Azure subscription ID.
+- `resource-group` - *(Optional)* The Azure [resource group](https://azure.microsoft.com/en-us/documentation/articles/resource-group-portal/) name of the cluster. You can find this in the Azure [portal](https://portal.azure.com) or [CLI](https://npmjs.org/azure-cli).
+- `scale-set` - *(Optional)* The Azure [VM scale set](https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/overview) name of the cluster. If this setting is configured, the plugin will search for instances over the resources only within this scale set.
+- `tag` - *(Optional)* The key-value pair of the tag on the Hazelcast vm resources. The format should be as `key=value`.
```
Do we mention about portRange?

alparslanavci replied:
Added 👍

gokhanoner commented on pom.xml:
```diff
@@ -62,301 +65,147 @@
+        <developer>
+            <id>gokhanoner</id>
+            <name>Gokhan Oner</name>
+            <email>gokhan@hazelcast.com</email>
```
@alparslanavci can you update my email to gokhanoner@gmail.com  & add a url section as https://github.com/gokhanoner ?

alparslanavci replied:
sure @gokhanoner, will push it directly since this PR is merged.

alparslanavci replied:
updated -> https://github.com/hazelcast/hazelcast-azure/blob/master/pom.xml#L69
